### ISSUE TITLE
find-file complete environment variables

### ISSUE THREAD

**Issue by Henry** [abo-abo/swiper#776]
With ivy and counsel-find-file is it possible to complete environment variables?

I have moved from icicles (which support environment variable completion) to ivy and miss this very useful feature.  With ido it is possible to work-around the limitation of no environment variable completion by reverting to the standard completion mechanism with C-f, is this possible with ivy if completing environment variables is not currently possible?

**Comment by abo-abo** [comment]
Can you explain a use case? There is already a workflow for starting file completion from a bookmark. Maybe that's what you want.

**Comment by Henry** [comment]
I have a large project which is split into source, documentation, libraries, tutorials _etc._ each of which has a corresponding environment variable for easy access to the files,  `cd` to the directories _etc._.  For example with icicles I can C-x C-f then $PRO<tab> -> $PROJECT_ then type T<tab> which completes to $PROJECT_TUTORIALS which then expands to the tutorials directory to which I add the rest of the path to the file.

The advantage of using environment variables rather than emacs-specific functionality is that they are available in the shell, shell scripts and other environment-aware utilities.

**Comment by abo-abo** [comment]
Here's a sketch of a solution:

```elisp
(setq my-env
      (mapcar (lambda (s) (split-string s "=" t))
              (split-string (shell-command-to-string "env") "\n" t)))

(defun maybe-expand-env ()
  (interactive)
  (if (equal ivy-text "")
      (ivy-read "var: " my-env
                :action (lambda (x) (insert (cadr x))))
    (insert "$")))

(define-key counsel-find-file-map (kbd "$") 'maybe-expand-env)
```

**Comment by Henry** [comment]
Thanks for coming-up with a possible solution so fast!.  I eval'd your sketch and typed C-x C-f then `$` and I get the error message:

ivy-read: Command attempted to use minibuffer while in minibuffer

**Comment by abo-abo** [comment]
> ivy-read: Command attempted to use minibuffer while in minibuffer

This is a generic Emacs error. Fixed by setting:

    (setq enable-recursive-minibuffers t)

Many people have the above setting in their config.

**Comment by Henry** [comment]
Perfect!!!

**Comment by Henry** [comment]
Do you think it is better storing the environment in `counsel-env` or filtering `process-environment`?  I am testing

    (defun counsel-env ()
          (mapcar (lambda (s) (split-string s "=" t))
                  process-environment))

    (defun counsel-expand-env ()
      (interactive)
      (if (equal ivy-text "")
          (ivy-read "var: " (counsel-env)
                    :action (lambda (x) (insert (cadr x))))
        (insert "$")))

which will pick-up any changes in the environment.  Do you see any problems with this approach?

**Comment by abo-abo** [comment]
Looks good, maybe remove non-existing directories like this:

```elisp
(defun counsel-env ()
  (delq nil
        (mapcar (lambda (s)
                  (let* ((res (split-string s "=" t))
                         (path (cadr res)))
                    (and
                     (stringp path)
                     (file-exists-p path)
                     res)))
                process-environment)))
```

**Comment by Henry** [comment]
This does what you would want nearly all the time, and is certainly the best approach for environment variables which are absolute paths.  However if an environment variable relates to a file name which is not in the current directory then `(file-exists-p path)` is false and the environment variable does not expand.  To get this to work the `(file-exists-p path)` would need to check the `path` relative to the currently selected path given to the `ivy-read` of the directory before `$` is pressed; is this possible?

**Comment by abo-abo** [comment]
> To get this to work the (file-exists-p path) would need to check the path relative to the currently selected path given to the ivy-read of the directory before $ is pressed; is this possible?

    (or (file-exists-p path)
        (file-exists-p (expand-file-name path ivy--directory)))

**Comment by Henry** [comment]
That nearly works in that the environment variable expands as it should but the edited path is lost.  For example I am in directory `~/.emacs.d/init` and press C-x C-f for which I get

    Find file: ~/.emacs.d/init

I backspace over the `init` leaving 

    Find file: ~/.emacs.d/init

then type `$` and select the environment variable 'HMM' which is set to `my-lisp` so it should complete to `~/.emacs.d/my-lisp` but instead it reinstates the `init` directory I backspaced over and completes to 

    Find file: ~/.emacs.d/init/my-lisp

which does not exist.

There is another potential issue if the environment variable refers to a file which does not exist and the intention is to create it.

**Comment by Henry** [comment]
I have tried several configurations but still having problems with the edited path being reinstated when `$` is pressed.  Is there a function which can be called when `$` is pressed to start from the currently displayed path rather than the unedited path being reinstated?

**Comment by abo-abo** [comment]
Can you provide more info. Like a set of reproduction steps?

**Comment by Henry** [comment]
Which part of the reproduction steps above is unclear and I will provide more detail.

**Comment by abo-abo** [comment]
Write for me e.g.:

    (setenv "HMM" "???")

**Comment by Henry** [comment]
I set the environment variable outside emacs before starting emacs which picks it up

    export HMM=my-lisp

but within emacs

    (setenv "HMM" "my-lisp")

should also work.

**Comment by Henry** [comment]
I have worked-around the problem to some extent.  To avoid the issue of the original path in the minibuffer being reinstated after the completion of the environment variable `ivy-read` I now expand the path of the environment variable completion candidates:

    (defun counsel-env ()
      (delq nil
            (mapcar
             (lambda (s)
               (let* ((res (split-string s "=" t))
                      (path (cadr res)))
                 (when (stringp path)
                   (cond ((file-exists-p path)
                          res)
                         ((file-exists-p (expand-file-name path ivy--directory))
                          (list (car res) (expand-file-name path ivy--directory)))
                         (t nil)))))
             process-environment)))

which works and completes correctly eventhough in the minibuffer I see

    Find file:  init//home/dm2/henry/.emacs.d/my-lisp

rather than

    Find file:  /home/dm2/henry/.emacs.d/my-lisp

so there is still the residual effect of the now spurious `init` directory which had been back-spaced over in the original `find-file` minibuffer but it doesn't affect the completion as pressing return selects `/home/dm2/henry/.emacs.d/my-lisp` as it should.

If the original `init` which had been back-spaced over could really be removed before the environment variable completion is started the expansion of the path in `counsel-env` would probably not be needed but I have tried lots of combinations of manipulating `ivy--current`, `ivy--directory` etc. with no luck.

**Comment by Henry** [comment]
I have done some more digging on handling recursive minibuffers, in particular `ivy-reverse-i-search ` and came up with this which appears to work exactly as intended:

    (defun counsel-env ()
      (delq nil
            (mapcar
             (lambda (s)
               (let* ((res (split-string s "=" t))
                      (path (cadr res)))
                 (when (stringp path)
                   (cond ((file-exists-p path)
                          res)
                         ((file-exists-p (expand-file-name path ivy--directory))
                          (list (car res) (expand-file-name path ivy--directory)))
                         (t nil)))))
             process-environment)))

    (defun counsel-expand-env ()
      (interactive)
      (if (equal ivy-text "")
          (progn
            (let ((enable-recursive-minibuffers t)
                  (history (symbol-value (ivy-state-history ivy-last)))
                  (old-last ivy-last)
                  (ivy-recursive-restore nil))
              (ivy-read "Env: " (counsel-env)
                        :action (lambda (x)
                                  (ivy--reset-state (setq ivy-last old-last))
                                  (setq ivy--directory "")
                                  (delete-minibuffer-contents)
                                  (insert (cadr x))))))
        (insert "$")))

however, I am sure it could be simplified.

**Comment by abo-abo** [comment]
Doesn't look too bad, but sure, it should be simpler. I'll have to take a closer look when I get some time.

**Comment by Henry** [comment]
I have added some functionality to abbreviate the path resulting from the environment variable expansion and added a trailing `/` if it is a directory so that the original `find-file` `ivy-read` treats the result more consistently for further completion:

    (defun counsel-env-res (res path)
      (let ((apath (-file-name path)))
        (list (car res)
              (if (file-accessible-directory-p path)
                  (file-name-as-directory apath)
                apath))))

    (defun counsel-env ()
      (delq nil
            (mapcar
             (lambda (s)
               (let* ((res (split-string s "=" t))
                      (path (cadr res)))
                 (when (stringp path)
                   (cond ((file-exists-p path)
                          (counsel-env-res res path))
                         ((file-exists-p (expand-file-name path ivy--directory))
                          (counsel-env-res
                           res (expand-file-name path ivy--directory)))
                         (t nil)))))
             process-environment)))

    (defun counsel-expand-env ()
      (interactive)
      (if (equal ivy-text "")
          (progn
            (let ((enable-recursive-minibuffers t)
                  (history (symbol-value (ivy-state-history ivy-last)))
                  (old-last ivy-last)
                  (ivy-recursive-restore nil))
              (ivy-read "Env: " (counsel-env)
                        :action (lambda (x)
                                  (ivy--reset-state (setq ivy-last old-last))
                                  (setq ivy--directory "")
                                  (delete-minibuffer-contents)
                                  (insert (cadr x))))))
        (insert "$")))

Now it works really well.

**Comment by TauPan** [comment]
Hey! Thanks to you guys I now have the following in my `.emacs` which works really well:
```emacs-lisp
        ;; from https://github.com/abo-abo/swiper/issues/776#issuecomment-263833155
        (defun counsel-env-res (res path)
          (let ((apath (abbreviate-file-name path)))
            (list (car res)
                  (if (file-accessible-directory-p path)
                      (file-name-as-directory apath)
                    apath))))

        (defun counsel-env ()
          (delq nil
                (mapcar
                 (lambda (s)
                   (let* ((res (split-string s "=" t))
                          (path (cadr res)))
                     (when (stringp path)
                       (cond ((file-exists-p path)
                              (counsel-env-res res path))
                             ((file-exists-p (expand-file-name path ivy--directory))
                              (counsel-env-res
                               res (expand-file-name path ivy--directory)))
                             (t nil)))))
                 process-environment)))

        (defun counsel-expand-env ()
          (interactive)
          (if (equal ivy-text "")
              (progn
                (let ((enable-recursive-minibuffers t)
                      (history (symbol-value (ivy-state-history ivy-last)))
                      (old-last ivy-last)
                      (ivy-recursive-restore nil))
                  (ivy-read "Env: " (counsel-env)
                            :action (lambda (x)
                                      (ivy--reset-state (setq ivy-last old-last))
                                      (setq ivy--directory "")
                                      (delete-minibuffer-contents)
                                      (insert (cadr x))))))
            (insert "$")))
        (define-key counsel-find-file-map (kbd "$") 'counsel-expand-env)
```

Any plans on making this permanent (i.e. closing this issue)? ;)

(My use-case is only slightly different from @Henry 's as I keep zsh named path aliases for projects which I use for fast switching in zsh via `~NAME` and so far I could only access those in emacs via `$`. I'm seriously considering switching the activation char to `~` but that would maybe conflict with expansion of user's home directories.)

**Comment by abo-abo** [comment]
> Any plans on making this permanent (i.e. closing this issue)? ;)

@TauPan if you or @Henry could make a pull request that would move things along faster. I don't use environment vars in my workflow, so it's low priority for me.

**Comment by tangxinfa** [comment]
@Henry 's solution works good.

But `counsel-find-file` has too many problems for works intuitively:

- path with environment variables

- tramp path default to ssh

- flexible edit path

  We also use `file-file` to create new file, especially painful when create
  file based on exists file name or directory name.

Finally i decide to use default find-file:

    (define-key counsel-mode-map [remap find-file] #'find-file)
    (add-to-list 'ivy-completing-read-handlers-alist '(find-file . completing-read-default))

Hopes there is a built-in option to disable `counsel-find-file`.

**Comment by abo-abo** [comment]
> Hopes there is a built-in option to disable counsel-find-file.

Your two lines of code is the built-in option to disable it.
